File: core/blas/dot.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include "dot.h"
#include "blas_common.h"
#include "blasplan/include/blasplan/BlasDotPlan.h"

using namespace AsdSip;
using namespace Mki;

constexpr int ELEMENTS_EACH_COMPLEX64 = 2;

namespace AsdSip {

struct DotImplParam {
    int64_t n;
    aclTensor *x;
    int64_t incx;
    aclTensor *y;
    int64_t incy;
    aclTensor *result;
    int64_t isConj;
};

AspbStatus asdBlasDotImpl(asdBlasHandle handle, DotImplParam implParam, int64_t sdot)
{
    ASDSIP_ECHECK(BlasPlanCache::doesPlanExist(handle),
        "blas Dot get cached plan failed.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

    int64_t *storageDims = nullptr;
    uint64_t dotStorageDimsNum = 0;
    int64_t checkSize = 0;
    CHECK_STATUS_WITH_ACL_RETURN(aclGetStorageShape(implParam.x, &storageDims, &dotStorageDimsNum),
        "asdBlasDotImpl: aclGetStorageShape");
    checkSize = sdot == 1 ? *storageDims : *storageDims * ELEMENTS_EACH_COMPLEX64;
    if (checkSize != implParam.n) {
        delete[] storageDims;
        storageDims = nullptr;
        ASDSIP_ELOG(ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH) << "blas asdBlasDotImpl get wrong x tensor num.";
        return ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH;
    }

    delete[] storageDims;
    storageDims = nullptr;

    CHECK_STATUS_WITH_ACL_RETURN(aclGetStorageShape(implParam.y, &storageDims, &dotStorageDimsNum),
        "asdBlasDotImpl: aclGetStorageShape");
    checkSize = sdot == 1 ? *storageDims : *storageDims * ELEMENTS_EACH_COMPLEX64;
    if (checkSize != implParam.n) {
        delete[] storageDims;
        storageDims = nullptr;
        ASDSIP_ELOG(ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH) << "blas asdBlasDotImpl get wrong y tensor num.";
        return ErrorType::ACL_ERROR_OP_INPUT_NOT_MATCH;
    }

    if (storageDims != nullptr) {
        delete[] storageDims;
        storageDims = nullptr;
    }

    ASDSIP_ECHECK(implParam.n > 0, "blas asdBlasDotImpl get wrong input n, please check.",
        ErrorType::ACL_ERROR_INVALID_PARAM);

    if (implParam.incx <= 0) {
        ASDSIP_LOG(INFO) << "blas asdBlasDotImpl get incx <= 0,change to 1.";
        implParam.incx = 1;
    }

    if (implParam.incy <= 0) {
        ASDSIP_LOG(INFO) << "blas asdBlasDotImpl get incy <= 0,change to 1.";
        implParam.incy = 1;
    }

    Status status;
    OpDesc opDesc;
    opDesc.opName = "DotOperation";
    AsdSip::OpParam::Dot param;
    param.n = implParam.n;
    param.isConj = implParam.isConj;
    opDesc.specificParam = param;
    ASDSIP_LOG(DEBUG) << "OpDesc: " << opDesc.opName << "; OpDesc info: " << param.ToString();

    SVector<aclTensor *> dotInTensors{implParam.x, implParam.y};
    SVector<aclTensor *> dotOutTensors{implParam.result};

    try {
        AsdSip::BlasDotPlan &plan = dynamic_cast<AsdSip::BlasDotPlan &>(BlasPlanCache::getPlan(handle));
        ASDSIP_ECHECK(plan.IsInitialized(), "Dot plan init Error!.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

        status = RunAsdOpsV2(plan.GetStream(), opDesc, dotInTensors, dotOutTensors, plan.GetWorkspace());
        ASDSIP_ECHECK(status.Ok(), status.Message(), ErrorType::ACL_ERROR_INTERNAL_ERROR);

        ASDSIP_LOG(INFO) << "Execute asdBlasDot success.";
        return ErrorType::ACL_SUCCESS;
    } catch (std::bad_cast &e) {
        // Handle the op dot exception
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Dot Error: " << e.what();
        return ErrorType::ACL_ERROR_INTERNAL_ERROR;
    }
}

AspbStatus asdBlasSdot(asdBlasHandle handle, const int64_t n, aclTensor *x, const int64_t incx, aclTensor *y,
                       const int64_t incy, aclTensor *result)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    aclDataType dataType = aclDataType::ACL_DT_UNDEFINED;
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(x, &dataType), "asdBlasSdot: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_FLOAT, "blas asdBlasSdot get wrong x tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);

    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(y, &dataType), "asdBlasSdot: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_FLOAT, "blas asdBlasSdot get wrong y tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(result, &dataType), "asdBlasSdot: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_FLOAT, "blas asdBlasSdot get wrong result tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    DotImplParam param = {n, x, incx, y, incy, result, 0};
    return asdBlasDotImpl(handle, param, 1);
}

AspbStatus asdBlasCdotu(asdBlasHandle handle, const int64_t n, aclTensor *x, const int64_t incx, aclTensor *y,
                        const int64_t incy, aclTensor *result)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    aclDataType dataType = aclDataType::ACL_DT_UNDEFINED;
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(x, &dataType), "asdBlasCdotu: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotu get wrong x tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);

    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(y, &dataType), "asdBlasCdotu: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotu get wrong y tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(result, &dataType), "asdBlasCdotu: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotu get wrong result tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    ASDSIP_ECHECK(n > 0 && n <= UINT32_MAX,
                  "blas asdBlasCdotu get n <= 0 or n > 2^32.", ErrorType::ACL_ERROR_INVALID_PARAM);
    DotImplParam param = {ELEMENTS_EACH_COMPLEX64 * n, x, incx, y, incy, result, 0};
    return asdBlasDotImpl(handle, param, 0);
}

AspbStatus asdBlasCdotc(asdBlasHandle handle, const int64_t n, aclTensor *x, const int64_t incx, aclTensor *y,
                        const int64_t incy, aclTensor *result)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    aclDataType dataType = aclDataType::ACL_DT_UNDEFINED;
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(x, &dataType), "asdBlasCdotc: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotc get wrong x tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);

    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(y, &dataType), "asdBlasCdotc: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotc get wrong y tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    CHECK_STATUS_WITH_ACL_RETURN(aclGetDataType(result, &dataType), "asdBlasCdotc: aclGetDataType");
    ASDSIP_ECHECK(dataType == aclDataType::ACL_COMPLEX64, "blas asdBlasCdotc get wrong result tensor dtype.",
                  ErrorType::ACL_ERROR_UNSUPPORTED_DATA_TYPE);
    ASDSIP_ECHECK(n > 0 && n <= UINT32_MAX,
                  "blas asdBlasCdotc get n <= 0 or n > 2^32.", ErrorType::ACL_ERROR_INVALID_PARAM);
    DotImplParam param = {ELEMENTS_EACH_COMPLEX64 * n, x, incx, y, incy, result, 1};
    return asdBlasDotImpl(handle, param, 0);
}

AspbStatus asdBlasMakeDotPlan(asdBlasHandle handle)
{
    std::lock_guard<std::mutex> lock(blas_mtx);
    ASDSIP_ECHECK(handle != nullptr, "blas MakeDotPlan Fail.", ErrorType::ACL_ERROR_INTERNAL_ERROR);

    AsdSip::BlasDotPlan *plan = nullptr;
    try {
        plan = new AsdSip::BlasDotPlan();
        BlasPlanCache::MakePlan(handle, plan);
    } catch (const std::exception& e) {
        if (plan != nullptr) {
            delete plan;
        }
        delete static_cast<int*>(handle);
        ASDSIP_ELOG(ErrorType::ACL_ERROR_INTERNAL_ERROR) << "Make Dot Plan failed: " << e.what();
        throw std::runtime_error("Make Dot Plan failed.");
    }
    plan->MarkInitialized();
    return ErrorType::ACL_SUCCESS;
}
}

File: sip_pta/test/blas/dot.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

import logging
import os
import sys
from dataclasses import dataclass
from typing import Optional

import torch
import torch_npu
import torch_sip

# 配置全局日志 (G.LOG.02)
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[logging.StreamHandler(sys.stdout)]
)
logger = logging.getLogger("torch_sip_dot_test")


@dataclass(frozen=True)
class DotConfig:
    """参数对象：封装向量内积的维度、类型及共轭配置"""
    n: int
    dtype: torch.dtype
    conjugate: bool = False
    name: str = "Case"


class DotTester:
    """向量内积 (Dot) 算子测试类"""

    def __init__(self, device: str = "npu:0"):
        self.device = device
        blocking_env = os.getenv("ASCEND_LAUNCH_BLOCKING", "0")
        mode_str = "异步 (Async)" if blocking_env == "0" else "同步 (Blocking)"
        logger.info("当前 NPU 模式: %s (BLOCKING=%s)", mode_str, blocking_env)

    @staticmethod
    def get_tensor(n: int, dtype: torch.dtype, device: str) -> torch.Tensor:
        """静态方法：安全生成测试 Tensor (先 CPU 后搬移)"""
        if dtype == torch.complex64:
            real = torch.randn(n, dtype=torch.float32)
            imag = torch.randn(n, dtype=torch.float32)
            return torch.complex(real, imag).to(device)

        return torch.randn(n, dtype=torch.float32).to(device)

    def run_case(self, cfg: DotConfig) -> bool:
        """
        执行单次测试用例。
        G.ERR.01: 最小化 try 块。
        """
        # 1. 准备数据
        vec_x = self.get_tensor(cfg.n, cfg.dtype, self.device)
        vec_y = self.get_tensor(cfg.n, cfg.dtype, self.device)

        # 2. 核心调用 (G.ERR.01: 仅包裹算子执行及其同步)
        try:
            res_npu = torch_sip.asd_blas_dot(vec_x, vec_y, cfg.conjugate)
        except Exception as exc:
            logger.error("[%s] 算子执行崩溃: %s", cfg.name, exc)
            return False

        # 3. CPU 参考计算 (移出 try 块)
        x_cpu = vec_x.cpu()
        y_cpu = vec_y.cpu()

        if cfg.dtype == torch.float32:
            ref = torch.dot(x_cpu, y_cpu)
            mode_name = "SDOT"
        else:
            # 复数点积逻辑拆分
            if cfg.conjugate:
                # CDOTC 定义: dotc(x, y) = sum(conj(x_i) * y_i)
                ref = (x_cpu.conj() * y_cpu).sum()
                mode_name = "CDOTC"
            else:
                # CDOTU 定义: dotu(x, y) = sum(x_i * y_i)
                ref = (x_cpu * y_cpu).sum()
                mode_name = "CDOTU"

        # 4. 验证结果
        is_close = torch.allclose(res_npu.cpu(), ref, rtol=1e-3, atol=1e-3)

        status = "PASS" if is_close else "FAIL"
        logger.info("[%s] %-20s | %s | Conj=%s | N=%d",
                    status, cfg.name, mode_name, cfg.conjugate, cfg.n)

        if not is_close:
            logger.error("      Max Diff: %.6e", torch.abs(res_npu.cpu() - ref).max())

        return is_close


def main():
    """测试主入口"""
    tester = DotTester()
    logger.info("开始 Dot (SDOT/CDOTU/CDOTC) 专项测试 ...\n")

    # 使用参数对象定义测试套件
    test_suites = [
        DotConfig(1024, torch.float32, name="Float_Dot"),
        DotConfig(2048, torch.complex64, conjugate=False, name="Complex_Dotu"),
        DotConfig(2048, torch.complex64, conjugate=True, name="Complex_Dotc"),
    ]

    all_passed = True
    for config in test_suites:
        if not tester.run_case(config):
            all_passed = False

    logger.info("-" * 60)
    if all_passed:
        logger.info("测试结论: ✅ 全部通过")
    else:
        logger.error("测试结论: ❌ 存在失败项")

    return 0 if all_passed else 1


if __name__ == "__main__":
    sys.exit(main())